polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -2901,3 +2903,76 @@ ValueCategory MLIRScanner::EmitBinOr(const BinOpInfo &Info) {
                                              Info.getRHS().getValue(builder)),
                        /*isReference*/ false);
 }
+
+#define HANDLEUNARYOP(OP)                                                      \
+  ValueCategory MLIRScanner::VisitUnary##OP(UnaryOperator *E,                  \
+                                            QualType PromotionType) {          \
+    LLVM_DEBUG({                                                               \
+      llvm::dbgs() << "VisitUnary" #OP ": ";                                   \
+      E->dump();                                                               \
+      llvm::dbgs() << "\n";                                                    \
+    });                                                                        \
+    QualType promotionTy =                                                     \
+        PromotionType.isNull()                                                 \
+            ? Glob.getTypes().getPromotionType(E->getSubExpr()->getType())     \
+            : PromotionType;                                                   \
+    auto result = Visit##OP(E, promotionTy);                                   \
+    if (result.val && !promotionTy.isNull())                                   \
+      result = EmitUnPromotedValue(getMLIRLocation(E->getExprLoc()), result,   \
+                                   E->getType());                              \
+    return result;                                                             \
+  }
+#include "Expressions.def"
+
+ValueCategory MLIRScanner::VisitPlus(UnaryOperator *E, QualType PromotionType) {
+  if (!PromotionType.isNull())
+    return EmitPromotedScalarExpr(E->getSubExpr(), PromotionType);
+  return Visit(E->getSubExpr());
+}
+
+ValueCategory MLIRScanner::VisitMinus(UnaryOperator *E,
+                                      QualType PromotionType) {
+  const auto Loc = getMLIRLocation(E->getExprLoc());
+  ValueCategory Op;
+  if (!PromotionType.isNull())
+    Op = EmitPromotedScalarExpr(E->getSubExpr(), PromotionType);
+  else
+    Op = Visit(E->getSubExpr());
+
+  // Generate a unary FNeg for FP ops.
+  if (mlirclang::isFPOrFPVectorTy(Op.val.getType()))
+    return Op.FNeg(builder, Loc);
+
+  // Emit unary minus with EmitBinSub so we handle overflow cases etc.
+  const auto Zero = ValueCategory::getNullValue(builder, Loc, Op.val.getType());
+  return EmitBinSub(
+      BinOpInfo{Zero, Op, E->getType(), BinaryOperator::Opcode::BO_Sub, E});
+}
+
+ValueCategory MLIRScanner::VisitImag(UnaryOperator *E, QualType PromotionType) {
+  Expr *Op = E->getSubExpr();
+
+  assert(!Op->getType()->isAnyComplexType() && "Unsupported");
```

Comment:
A bit confusing, the member function name `VisitImag` made me think we would receive an expression that has Comples type and just visit the immaginary subexpression. Can you explain in a comment perhaps the intention of this function?

Reply:
We can receive any scalar expression AFAIC here. `__imag__(X) = 0` for every scalar `X`.

Reply:
There is already a comment below explaining the behavior for scalar input expressions.